prov/rxm/src/rxm_cq.c

```diff
@@ -409,25 +409,24 @@ ssize_t rxm_cq_handle_data(struct rxm_rx_buf *rx_buf)
 
 static ssize_t rxm_handle_recv_comp(struct rxm_rx_buf *rx_buf)
 {
-	struct rxm_recv_match_attr match_attr;
+	struct rxm_recv_match_attr match_attr = {
+		.addr = FI_ADDR_UNSPEC,
+	};
 	struct dlist_entry *entry;
-	struct util_cq *util_cq;
-
-	util_cq = rx_buf->ep->util_ep.rx_cq;
 
-	if ((rx_buf->ep->rxm_info->caps & (FI_SOURCE | FI_DIRECTED_RECV)) &&
-	    !rx_buf->conn) {
-		rx_buf->conn = rxm_key2conn(rx_buf->ep, rx_buf->pkt.ctrl_hdr.conn_id);
-		if (OFI_UNLIKELY(!rx_buf->conn))
-			return -FI_EOTHER;
-		match_attr.addr = rx_buf->conn->handle.fi_addr;
-	} else {
-		match_attr.addr = FI_ADDR_UNSPEC;
+	if (rx_buf->ep->rxm_info->caps & FI_SOURCE) {
+		if ((rx_buf->ep->rxm_info->caps & FI_DIRECTED_RECV) && !rx_buf->conn) {
+			rx_buf->conn = rxm_key2conn(rx_buf->ep,
+						    rx_buf->pkt.ctrl_hdr.conn_id);
+			if (OFI_UNLIKELY(!rx_buf->conn))
+				return -FI_EOTHER;
+			match_attr.addr = rx_buf->conn->handle.fi_addr;
+		}
+		rx_buf->ep->util_ep.rx_cq->src[
+			ofi_cirque_windex(rx_buf->ep->util_ep.rx_cq->cirq)] =
+				match_attr.addr;
```

Comment:
> I update the PR to use match_attr.addr instead of rx_buf->conn->handle.fi_addr
Does it make sense?

Still, match_attr.addr is set only when FI_DIRECTED_RECV is set. This change breaks existing behavior. We need to set match_attr.addr (thereby rx_cq->src[...]) whenever FI_SOURCE or FI_DIRECTED_RECV is requested.

Reply:
> Still, match_attr.addr is set only when FI_DIRECTED_RECV is set.

Sorry, it seems that the logic is the same as previous:
- `match_addr.addr` is set when `FI_SOURCE` and `FI_DIRECTED_RECV` are provided
- r`x_cq->src[...]` is set when `FI_SOURCE` is provided at least.

Reply:
Yes, but previously `rx_cq->src[...]` was being set to `rx_buf->conn->handle.fi_addr` which in turn would be set if either `FI_SOURCE` or `FI_DIRECTED_RECV` is set.

Reply:
Now I see! I missed this. Thanks.

Reply:
@a-ilango I've updated the PR with new changes.
Looking at the code, `rx_buf->conn` is always `NULL`. `rxm_ep_repost_buf` sets it to `NULL` and issues `fi_recv()`.
When the data (type of message is `ofi_ctrl_data`) is arriving into this buffer and CQ is completed, the function `rxm_handle_recv_comp` is called to process this buffer.
So, there are no needs to check whether the `rx_buf->conn == NULL`, we can just  call `rxm_key2conn` to retrieve connection object.

Reply:
That makes sense. We can skip the NULL check for rx_buf->conn.